lib/hipchat/errors.rb

```diff
@@ -1,14 +1,49 @@
 module HipChat
-  class ServiceError        < StandardError; end
-  class UnknownRoom         < ServiceError; end
-  class RoomNameTooLong     < ServiceError; end
-  class RoomMissingOwnerUserId < ServiceError; end
-  class Unauthorized        < ServiceError; end
-  class UsernameTooLong     < ServiceError; end
-  class UnknownResponseCode < ServiceError; end
-  class UnknownUser         < ServiceError; end
-  class InvalidApiVersion   < ServiceError; end
-  class InvalidUrl          < ServiceError; end
-  class InvalidEvent        < ServiceError; end
-  class UnknownWebhook      < ServiceError; end
+  class ServiceError            < StandardError;       end
+  class RoomNameTooLong         < ServiceError;        end
+  class RoomMissingOwnerUserId  < ServiceError;        end
+  class UnknownResponseCode     < ServiceError;        end
+  class Unauthorized            < UnknownResponseCode; end
+  class BadRequest              < ServiceError;        end
+  class MethodNotAllowed        < ServiceError;        end
+  class UsernameTooLong         < ServiceError;        end
+  class InvalidApiVersion       < ServiceError;        end
+  class InvalidUrl              < ServiceError;        end
+  class InvalidEvent            < ServiceError;        end
+  class ObjectNotFound          < UnknownResponseCode; end
+  class UnknownRoom             < ObjectNotFound;      end
+  class UnknownUser             < ObjectNotFound;      end
+  class UnknownWebhook          < ObjectNotFound;      end
+  class TooManyRequests         < UnknownResponseCode; end
+
+  class ErrorHandler
+
+    # Pass-through to catch error states and raise their appropriate exceptions
+    # @param klass {Symbol} The class of object we are handling the error for
+    # @param identifier {String} An identifying string for the object causing the error
+    # @param response {HTTParty::Response} The HTTParty response/request object
+    def self.response_code_to_exception_for(klass, identifier, response)
+      # Supports user, room and webhook objects.  If we get something other than that, bail.
+      raise ServiceError "Unknown class #{klass}" unless [:user, :room, :webhook].include? klass
+      # Grab the corresponding unknown object exception class and constantize it for the 404 case to call
+      not_found_exception = Module.const_get(HipChat.to_s.to_sym).const_get("Unknown#{klass.capitalize}".to_sym)
```

Comment:
What would happen with the scenario where the class doesn't have a UnknownClass? Could you default to ServiceError?